### PR TITLE
Binary search oracle

### PR DESCRIPTION
**Pull request by shromonag**
This is the pseudo code, need to test though.

Have to add in the call for the SMT.

### CODE REVIEW

File: magnum/solvers/search_oracle.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+def check_sat(w_star, u_star, r, g, num_samples=100):
+    A, B, C = g.model.dyn
+    dt = g.model.dt
+    # Adjust for discrete time
+    A = np.eye(len(g.model.vars.state)) + dt * A
+    B = dt * B
+    C = dt * C
+
+    times = (g.model.dt * t for t in g.times)
```
Comment (mvcisback) [inline magnum/solvers/search_oracle.py]: I justed added some code in `game.py` for scaled times. 

`times = g.scaled_times`

File: magnum/solvers/search_oracle.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    phi = g.spec_as_stl(dizcretize=False)
+    phi_eval = stl.boolean_eval.pointwise_sat(phi)
+
+    for _ in range(num_samples):
+        controls = {name: random_series(g) * r + u_star[name] for name in
+                g.model.vars.input}
+        iterate_t = lambda t: extract_t(controls=controls, w_star=w_star, t=t)
+        x_ts = reduce(lambda x, cont: A * x[-1] + B * cont['u'] + C(cont['w']),
+                  [iterate_t in g.times], [x0])
```
Comment (mvcisback) [inline magnum/solvers/search_oracle.py]: do you mean `map(iterate_t, g.times)`?

File: magnum/solvers/search_oracle.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+
+    while r_high - r_low > epsilon:
+        if r_high <= r_low:
+            break
+        if check_sat(w_star, u_star, r_mid, g, num_samples):
+            r_low = r_mid
+        else:
+            r_high = r_mid
+        r_mid = (r_low + r_high) / 2.
+
```
Comment (mvcisback) [inline magnum/solvers/search_oracle.py]: Shouldn't this have an outer loop that performs an smt query to assert that everything inside the loop doesn't satisify the spec?

File: magnum/solvers/search_oracle.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+import stl
+import numpy as np
+
+import traces
+from functools import reduce
+
+def random_series(g):
+    times = (g.model.dt*t for t in g.times)
+    return traces.TimeSeries(zip(times, np.random.random(len(g.times))))
```
Comment (mvcisback) [inline magnum/solvers/search_oracle.py]: Shouldn't this depend on the radius and u?

**Review by mvcisback — CHANGES_REQUESTED** [review]
Could you add a test to `test_cegis.py` that tests that the region you are expecting is recoved. For example in rock paper scissors, if the adversary plays rock you should learn not paper.

File: magnum/solvers/search_oracle.py (hunk 4 above)
Comment (shromonag) [reply magnum/solvers/search_oracle.py]: No  I have taken care of that when I assign 'u'

File: magnum/solvers/search_oracle.py (hunk 3 above)
Comment (shromonag) [reply magnum/solvers/search_oracle.py]: Yes I have to add in the SMT query, this is just the largest square finding thing. I want to test this first.

File: magnum/solvers/search_oracle.py (hunk 4 above)
Comment (mvcisback) [reply magnum/solvers/search_oracle.py]: ah I see that now

**Review by mvcisback — CHANGES_REQUESTED** [review]

**Review by mvcisback — APPROVED** [review]

**Pull request merged**